### PR TITLE
Fix restoring trainer after `lr_find()`

### LINKED ISSUE #12670: Trainer always causes divergence after `Trainer.tuner.lr_find`
## 🐛 Bug

We determine the learning rate using Lightnings LR finder as follows:

```python
trainer = pl.Trainer(max_epochs=16, gpus=1, auto_lr_find=True)
lr_finder = trainer.tuner.lr_find(
    model, train_dataloaders=train_loader,
)
lr = lr_finder.suggestion() or 0.001
```

Then, we re-instantiate the module with the new learning rate and call `trainer.fit()` _using the same Trainer instance_ as for finding the learning rate.

In PyTorch Lightning 1.5.x, this worked without issues. In Lightning 1.6, this always leads to divergence, regardless of the learning rate. Using a new Trainer instance works, but from my understanding, the trainer should still be usable after having called `trainer.tuner.lr_find()`.

### To Reproduce

The following Colab reproduces the issue with the Boring Model: https://colab.research.google.com/drive/12BF8m4bCe6ijWDQ1KCEXr5wNAL0TCyhW?usp=sharing

Changing the pytorch_lightning version in the first cell to 1.5.10 solves the issue.

### Expected behavior

Calling `trainer.fit()` after `trainer.tuner.lr_find()` works just like calling `trainer.fit()` right after instantiating the trainer.

### Environment

* CUDA:
	- GPU:
		- Tesla K80
	- available:         True
	- version:           11.1
* Packages:
	- numpy:             1.21.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.10.0+cu111
	- pytorch-lightning: 1.6.0
	- tqdm:              4.63.0
* System:
	- OS:                Linux
	- architecture: 64bit
	- processor:         x86_64
	- python:            3.7.13

cc @akihironitta @borda @rohitgr7

### PR DESCRIPTION
**Pull request by Callidior**
## What does this PR do?

Fixes #12670: After using the LR finder and then training the model with the same trainer instance (which is the standard use-case), an incorrect learning rate schedule is used, leading to immediate divergence.

This bug was introduced in Lightning version 1.6 by commit a6a28e08d22f59f2468ff1049c67507df7519e7c, which removed two lines responsible for resetting the `init_optimizers` method of the trainer after it has been replaced with a wrapper changing the LR scheduler to one trying different learning rates.

This PR reintroduces these two lines and updates them to match the current codebase, where `trainer.init_optimizers` has been renamed to `trainer.strategy.setup_optimizers`.

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/tuner/lr_finder.py [hunk 1]
```diff
@@ -297,6 +298,7 @@ def __lr_finder_restore_params(trainer: "pl.Trainer", params: Dict[str, Any]) ->
     trainer.callbacks = params["callbacks"]
     trainer.logger = params["logger"]
     trainer.fit_loop.max_steps = params["max_steps"]
+    trainer.strategy.setup_optimizers = params["setup_optimizers"]
```
Comment (carmocca) [inline src/pytorch_lightning/tuner/lr_finder.py:301]: Can you add a test to avoid regressions?

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/tuner/lr_finder.py (hunk 1 above)
Comment (Callidior) [reply src/pytorch_lightning/tuner/lr_finder.py:301]: Thanks for the suggestion. I've extended the existing test `test_trainer_reset_correctly` to also check the attribute `trainer.strategy.setup_optimizers`, which it wasn't doing so far.

**Review by carmocca — APPROVED** [review]
Great!

**Review by rohitgr7 — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
FYI, this PR created an issue with convergence and I am investigating now what the cause is. Repro here: #14824
